### PR TITLE
Implements Multiple Contacts per Body

### LINKED ISSUE #181: Finish implementing "multiple contacts per collision pair" for more accurate simulation
For example, the visualization of systems/plants/test/fallingBrickLCP shows the brick bouncing significantly off the ground. This behavior arose after the collision upgrade (a8609e66db1df43477631cd318f42f88e21f6b96 behaves normally).

### LINKED ISSUE #790: Numerical issues with bullet normals
Occasionally, bullet will return contact normals which are not actually normal to the contact surface.  

This is due to the fact that zero radius spheres are used to represent contact points.  Since the distance between the contact point and the center of the sphere can be extremely small, the normalization operation is numerically ill posed.  

To confirm this, I've set up an experiment that plots the error in the normals as a function of the radius of the contact spheres:

bulletnormals

As can be seen, as the radius of the contact sphere decreases below 1e-8, the error in the normals rises sharply and numerical artifacts begin to appear.

One way to address this would be to stop using zero radius spheres for contact points and instead give them a small but finite radius.  The non-bullet code that checks explicitly for zero radius spheres would also have to be changed to be tolerant of the new radius.

### PR DESCRIPTION
**Pull request by psiorx**
Initial functionality of multiple contact points per body is now working.  I've added a field to TSRBM to turn it on (off by default since its slower and not needed in all cases).  There's also two examples of it handling simulations that were previously impossible. See #181 for more discussion.  Resolves #181 and also resolves #790 (provided that urdfs define micron sized spheres instead of zero radius for contact points).

Here's a screenshot of one of the examples with Atlas balancing on a ledge with significant heel overhang. 

This is also the first example of using bullet collisions successfully in an Atlas simulation to my knowledge.

heel_overhang

### CODE REVIEW

File: examples/Atlas/urdf/atlas_convex_hull.urdf [hunk 1]
```diff
@@ -65,7 +65,7 @@
    <collision group="heel">
       <origin rpy="0 0 0" xyz="-0.082 0.0624435 -0.081119"/>
       <geometry>
-        <sphere radius="0"/>
+        <sphere radius="1e-7"/>
```
Comment (avalenzu) [inline examples/Atlas/urdf/atlas_convex_hull.urdf]: Why this change?

File: systems/plants/RigidBodyManipulator.cpp [hunk 2]
```diff
@@ -676,6 +676,45 @@ bool RigidBodyManipulator::collisionDetect( VectorXd& phi,
...
 }
 
+bool RigidBodyManipulator::potentialCollisions(VectorXd& phi,
+                                            MatrixXd& normal,
+                                            MatrixXd& xA,
+                                            MatrixXd& xB,
+                                            vector<int>& bodyA_idx,
+                                            vector<int>& bodyB_idx,
+                                            bool use_margins)
+{
```
Comment (tkoolen) [inline systems/plants/RigidBodyManipulator.cpp]: I would prefer calling potentialCollisionPoints outside potentialCollisions, passing in potential_collisions as an argument to the method and having potentialCollisions not return anything.

File: systems/plants/RigidBodySphere.m [hunk 3]
```diff
@@ -92,7 +92,7 @@ function writeWRLShape(obj,fp,td)
       % @param  obj - RigidBodySphere object
       % @retval pts - 3xm array of points on this geometry (in link frame) that
       %               can collide with the world.
-      if obj.radius == 0
+      if obj.radius <1e-6
```
Comment (tkoolen) [inline systems/plants/RigidBodySphere.m]: add a space after <

File: systems/plants/collision/BulletModel.cpp [hunk 4]
```diff
@@ -33,6 +33,8 @@ namespace DrakeCollision
   BulletCollisionWorldWrapper::BulletCollisionWorldWrapper()
     : bt_collision_configuration(), bt_collision_broadphase(), filter_callback()
   {
+    bt_collision_configuration.setConvexConvexMultipointIterations(8,8);
+    bt_collision_configuration.setPlaneConvexMultipointIterations(8,8);
```
Comment (tkoolen) [inline systems/plants/collision/BulletModel.cpp]: Magic number --> named constant?
also, spaces after commas.

**Comment by avalenzu** [comment]
Looks good to me, apart from the existing inline comments.

File: systems/plants/collision/BulletModel.cpp [hunk 5]
```diff
@@ -202,6 +204,67 @@ namespace DrakeCollision
     return id;
   }
 
+  bool BulletModel::potentialCollisionPoints(bool use_margins, vector<PointPair> & potential_collisions)
```
Comment (tkoolen) [inline systems/plants/collision/BulletModel.cpp]: Why not just create the vector in this function and return it? `potentialCollisions`, which calls this method, now creates the vector the line before `potentialCollisionPoints` is called so it's not like you're saving on allocations. Return value optimization will elide the copy. Returning a boolean that says whether the size is greater than zero is a little superfluous I'd say.

File: systems/plants/TimeSteppingRigidBodyManipulator.m [hunk 6]
```diff
@@ -40,13 +41,18 @@
...
       typecheck(manip,'RigidBodyManipulator');
       obj = obj@DrakeSystem(0,manip.getNumStates(),manip.getNumInputs(),manip.getNumOutputs(),manip.isDirectFeedthrough(),manip.isTI());
       obj.manip = manip;
       if isa(manip,'PlanarRigidBodyManipulator')
         obj.twoD = true;
       end
-      
+
+      if isfield(options, 'multiple_contacts')
+        obj.multiple_contacts = options.multiple_contacts == true;
```
Comment (avalenzu) [inline systems/plants/TimeSteppingRigidBodyManipulator.m]: Maybe use
```matlab
typecheck(options.multiple_contacts, 'logical');
obj.multiple_contacts = options.multiple_contacts;
```

File: systems/plants/collision/BulletModel.cpp [hunk 7]
```diff
@@ -202,6 +204,67 @@ namespace DrakeCollision
...
+        pointOnA_1 << toVector3d(pointOnElemA), 1.0;
+        pointOnB_1 << toVector3d(pointOnElemB), 1.0;
+
+        Vector3d pointOnA;
+        Vector3d pointOnB;
+
+        pointOnA << elements[idA]->getLocalTransform().topRows(3) * pointOnA_1;
+        pointOnB << elements[idB]->getLocalTransform().topRows(3) * pointOnB_1;
+
+        c.addSingleResult(idA, idB, pointOnA, pointOnB, toVector3d(normalOnB), static_cast<double>(pt.getDistance()) + marginA + marginB);
```
Comment (tkoolen) [inline systems/plants/collision/BulletModel.cpp]: code formatting in this file: you're using camelcase for variables; should be underscores. I'd also prefer camelcase, but it's in the style guide.

File: examples/Atlas/urdf/atlas_convex_hull.urdf (hunk 1 above)
Comment (psiorx) [inline examples/Atlas/urdf/atlas_convex_hull.urdf]: This is to allow bullet collision to work well with the existing contact spheres that I have to leave in place for the controller/planner. See #790

**Comment by psiorx** [comment]
Thanks for the review, guys.  The changes are in.

File: examples/Atlas/urdf/atlas_convex_hull.urdf (hunk 1 above)
Comment (RussTedrake) [inline examples/Atlas/urdf/atlas_convex_hull.urdf]: if we really think that one should never use zero radius spheres, then I would prefer to leave it as zero in the urdfs, and parse it to 1e-7 in the urdf loader (perhaps using warnOnce).

File: systems/plants/collisionDetectmex.cpp [hunk 8]
```diff
@@ -33,83 +30,65 @@ void mexFunction( int nlhs, mxArray *plhs[],int nrhs, const mxArray *prhs[] ) {
...
                               active_bodies_idx,active_group_names);
     } else {
-      model-> collisionDetect(dist, normals, ptsA, ptsB, bodyA_idx, bodyB_idx,
+      model->collisionDetect(dist, normals, ptsA, ptsB, bodyA_idx, bodyB_idx,
                               active_bodies_idx);
     }
   } else {
-    if (active_group_names.size() > 0) {
-      model-> collisionDetect(dist, normals, ptsA, ptsB, bodyA_idx, bodyB_idx,
+    const bool multiple_contacts = mxIsLogicalScalarTrue(allow_multiple_contacts);
```
Comment (RussTedrake) [inline systems/plants/collisionDetectmex.cpp]: is this doing sufficient argument checking?  what happens if allow_multiple_contacts is a double?  or is NULL?

File: systems/plants/collisionDetectmex.cpp [hunk 9]
```diff
@@ -33,83 +30,65 @@ void mexFunction( int nlhs, mxArray *plhs[],int nrhs, const mxArray *prhs[] ) {
...
-  // First get the list of body indices for which to compute distances
-  // DEBUG
-  //cout << "collisionDetectmex: Parse active_bodies START" << endl;
-  // END_DEBUG
+
   const mxArray* active_collision_options = prhs[2];
-  const mxArray* body_idx = mxGetField(active_collision_options,0,"body_idx");
+  const mxArray* allow_multiple_contacts = prhs[1];
+
+  const mxArray* body_idx = mxGetField(active_collision_options, 0, "body_idx");
```
Comment (RussTedrake) [inline systems/plants/collisionDetectmex.cpp]: what if active_collision_options is null (or nrhs < 3)

File: systems/plants/collisionDetectmex.cpp [hunk 10]
```diff
@@ -19,9 +19,6 @@ using namespace std;
  */
 
 void mexFunction( int nlhs, mxArray *plhs[],int nrhs, const mxArray *prhs[] ) {
-  //DEBUG
-  //cout << "collisionDetectmex: START" << endl;
-  //END_DEBUG
 
   if (nrhs < 1) {
     mexErrMsgIdAndTxt("Drake:collisionDetectmex:NotEnoughInputs","Usage collisionDetectmex(model_ptr)");
```
Comment (RussTedrake) [inline systems/plants/collisionDetectmex.cpp]: update this to describe the (optional?) active_collision_options and allow_multiple_contacts

File: systems/plants/collisionDetectmex.cpp (hunk 8 above)
Comment (psiorx) [inline systems/plants/collisionDetectmex.cpp]: Bad things.  I'm now checking the types of all arguments here.

**Comment by psiorx** [comment]
This segfault happened while building iris in drake-distro.

File: systems/plants/RigidBodyGeometry.m [hunk 11]
```diff
@@ -123,7 +123,7 @@
             size = parseParamString(model,robotnum,char(thisNode.getAttribute('size')));
             obj = RigidBodyBox(size(:));
           case 'sphere'
-            r = parseParamString(model,robotnum,char(thisNode.getAttribute('radius')));
+            r = min(1e-7, parseParamString(model,robotnum,char(thisNode.getAttribute('radius'))));
```
Comment (RussTedrake) [inline systems/plants/RigidBodyGeometry.m]: need to do this in the c++ parser, too.

File: systems/plants/collisionDetectmex.cpp (hunk 8 above)
Comment (RussTedrake) [inline systems/plants/collisionDetectmex.cpp]: did the argument checking get pushed?

Comment (psiorx) [inline systems/plants/collisionDetectmex.cpp]: Yes.  It's higher up in the file.

File: systems/plants/RigidBodyGeometry.m (hunk 11 above)
Comment (psiorx) [inline systems/plants/RigidBodyGeometry.m]: Good point.  I'm on it.
